### PR TITLE
Re-Implement the event loop to be able to be used in a while let loop

### PR DESCRIPTION
**Pull request by TheNeikos**
**Ready for review!**

This targets #123 

---

I'm mainly already submitting this to get some feedback. Please do check out the example if the interface changes are warranted, or if I could do it better? (For now the only change, except from losing the iterator, is that you have to pass the event into the draw functions.)

### CODE REVIEW

**Comment by bvssvni** [comment]
@TheNeikos You are awesome!

File: examples/hello_piston.rs [hunk 1]
```diff
@@ -3,21 +3,22 @@ extern crate piston_window;
 use piston_window::*;
 
 fn main() {
-    let window: PistonWindow = WindowSettings::new("Hello Piston!", [640, 480])
+    let mut events: WindowEvents = WindowEvents::new();
```
Comment (bvssvni) [inline examples/hello_piston.rs]: Could be `let mut events = window.events();`

File: src/lib.rs [hunk 2]
```diff
@@ -246,135 +217,38 @@ impl<T, W> PistonWindow<T, W>
     /// Changes application structure.
     #[inline(always)]
     pub fn app_by_value<U>(self, app: U) -> PistonWindow<U, W> {
```
Comment (bvssvni) [inline src/lib.rs]: Can remove `app_by_value`.

Comment (bvssvni) [inline src/lib.rs]: Can remove `app_by_value`.

File: src/lib.rs [hunk 3]
```diff
@@ -246,135 +217,38 @@ impl<T, W> PistonWindow<T, W>
...
-        self.app(Rc::new(RefCell::new(app)))
+        self.app(app)
     }
 
     /// Renders 2D graphics.
-    pub fn draw_2d<F>(&self, f: F) where
+    pub fn draw_2d<F>(&mut self, e: &Event, f: F) where
```
Comment (bvssvni) [inline src/lib.rs]: Use `e: &E` where `E: GenericEvent`

File: src/lib.rs [hunk 4]
```diff
@@ -246,135 +217,38 @@ impl<T, W> PistonWindow<T, W>
...
+                args.viewport(),
+                f
+            );
+            self.encoder.flush(&mut self.device);
         }
     }
 
     /// Renders 3D graphics.
-    pub fn draw_3d<F>(&self, f: F) where
+    pub fn draw_3d<F>(&mut self, e: &Event, f: F) where
```
Comment (bvssvni) [inline src/lib.rs]: Use `e: &E` where `E: GenericEvent`

File: src/lib.rs [hunk 5]
```diff
@@ -410,25 +284,5 @@ impl<T, W> AdvancedWindow for PistonWindow<T, W>
...
-    }
-
-    fn set_bench_mode(&mut self, enable: bool) {
-        self.events.borrow_mut().set_bench_mode(enable);
-    }
-}
-
 /// Creates a new empty application.
-pub fn empty_app() -> Rc<RefCell<()>> { Rc::new(RefCell::new(())) }
+pub fn empty_app() -> () { () }
```
Comment (bvssvni) [inline src/lib.rs]: Can remove `empty_app` and just use `()`.

**Comment by bvssvni** [comment]
Can also use `W` instead of `Rc<RefCell<W>>`.

`PistonWindow::new` should take `mut window: W`.

Remove these unused imports:

```
use std::cell::RefCell;
use std::rc::Rc;
use std::any::Any;
```

File: examples/hello_piston.rs [hunk 6]
```diff
@@ -3,21 +3,22 @@ extern crate piston_window;
...
+    let mut window: PistonWindow = WindowSettings::new("Hello Piston!", [640, 480])
         .exit_on_esc(true)
         .build()
         .unwrap_or_else(|e| { panic!("Failed to build PistonWindow: {}", e) });
     println!("Press any button to enter inner loop");
-    for e in window {
-        e.draw_2d(|_c, g| {
+    while let Some(e) =  events.next(&mut window) {
```
Comment (bvssvni) [inline examples/hello_piston.rs]: There is an extra space here.

**Comment by TheNeikos** [comment]
I also made the documentation build (but not run) as it was on `ignore` before.

**Comment by bvssvni** [comment]
Looks good! There are some minor things left, but I'll merge and fix them myself.

**Pull request merged**